Implement threading

Sonico98 opened the pull request:
Disabled by default. Allows the user to specify the amount of parallel uploads through a parameter `-p`, with values ranging from 1 to 16 (iBroadcast's limit hasn't been tested, we may need to lower the maximum value).
The progress bar is not working for multithreaded uploads.
Testing on my network (50Mbps upload) has shown an improvement of 30 seconds when uploading a folder containing 500MiB of files. Improvements could be bigger on faster networks and/or with more threads.
**4 parallel uploads:**
![image](https://user-images.githubusercontent.com/61394886/192755779-93fe05f4-9d6f-4c01-9ce0-5e87ebbd342f.png)
**Default upload (single threaded):**
![image](https://user-images.githubusercontent.com/61394886/192755796-9178547f-0cab-4c62-a8ea-3b355d6a12b5.png)

I'm unsure about the user interface, should we print the filenames by default, regardless of the verbose parameter? Should we enable threading by default with a sane value, like ```-p 4```?

Sonico98 committed:
Implement threading

Disabled by default. Allows the user to specify the amount of parallel uploads through a parameter `-p`, with values ranging from 1 to 16 (iBroadcast's limit hasn't been tested, we may need to lower the maximum value). The progress bar is not working for multithreaded uploads.

deseven commented:
@Sonico98 that's really good, thank you! I was thinking about doing it myself, but since I'm using this script as a cron job in the middle of the night, I decided that I don't really care how long it takes.

Some thoughts:

1. I think we should follow [MediaSync Lite](https://project.ibroadcast.com) defaults and limits, so it's 3 parallel uploads by default and 6 max. The `-p` flag should override the default value of 3 and `-p 0` or `-p 1` should limit uploading to a single thread. Makes sense?
2. `upload` can probably be removed completely, one unified function that uses 1 thread or more should be easier to maintain.
3. Please bump the version to 0.5 and update the readme.
4. Any ideas how to redo the progress bar?
5. Yeah, printing filenames by default makes sense, I think. This was in the original script. The `-s` key is used to print only the summary.

Thiago-DA committed:
Use ThreadPoolExecutor instead of threading

The previous queuing logic had a problem where we were checking only if
the first thread in the queue finished its tasks, ignoring completely if
other threads could have finished earlier. Using ThreadPoolExecutor
solves this issue as the queuing is handled by python itself, while also
resulting in less and cleaner code.

Sonico98 commented:
@deseven thanks for the input!
1. You are right, I'll change it accordingly.
2. The way it currently works is by passing a single file to the existing ```upload``` function, and executing each individual file upload as a thread. Doing it this has a big performance hit when using single thread, hence why I decided to keep the already existing functionality intact (just checking already uploaded files, takes around 20 seconds with the new method single threaded, but 4 seconds using the current method). Any suggestions on how this should be handled instead?
3. Will do.
4. Not really. One way to do it may be using curses, but I would like to try and avoid adding external dependencies. I'll give it a try later.
5. I forgot checking for that argument, thanks!

Edit: My bad, I didn't notice I pushed with another account.

Thiago-DA committed:
Add parallel uploads to README

Thiago-DA committed:
Bump version to 0.5

Thiago-DA committed:
Update defaults

Update default amount of parallel uploads to match those of MediaSync Lite

Thiago-DA committed:
Respect `-s` on parallel uploads

deseven commented:
Ok, let me test it myself (this weekend, hopefully) and I will also try to figure something out with the progress bar (worst case scenario I'll just wipe it out).

Sonico98 commented:
Found the problem. Turns out the slowdown comes from doing the MD5 checks during upload inside threads. Since the computation is kind of a heavy CPU task, running it inside a thread is not a great idea, instead it should be done with multiprocessing (maybe in the future?).
For the time being, moving said functionality to a separate function and calling it before starting the upload process solves the issue... for small directories. If we compute the MD5 for every file before starting any upload, then for large directories the user will have to wait until the whole dir has been computed.
Something I can think of is computing MD5 for the first N files in the list of files, storing the latest accessed index somewhere, uploading said N files and repeat the whole thing, continuing from the stored index position. I'm not sure if it's the best way to go though.

deseven commented:
I don't see a problem in making another step when the md5 calculation will be made. It's one time only anyway. And after that we can upload files quicker.

So the overall logic might be as follows:

1. Build the file list.
2. Prompt user if they want to upload found files.
3. Calculate checksums (non-threaded).
4. Upload files (threaded).
5. Display the summary.

Sonico98 commented:
What do you think about informing the user about the calculations for large directories, then? Regardless of `-v`being used or not.
If the file list is bigger than, say, 500, print a statement asking the user to be patient while the files are being processed. Otherwise it might be awkward seeing nothing happen for 10+ minutes after entering 'Upload'.

deseven commented:
Yeah, makes sense, just omit it with `-s`.

Commit:
Perform MD5 calculations before uploading

Calculating the MD5 hashes inside threads causes performance issues, so
we calculate them beforehand in a non-threaded way.

Commit:
Move code around

Sonico98 commented:
(sorry if I'm being dense, I just happen to come up with many questions)
When trying to find a sweet spot as to how many files should there be to inform the user that hashes are being calculated, I got this idea, what do you think? I still haven't tried getting the progress bar to work with threads though.
![image](https://user-images.githubusercontent.com/61394886/193161682-8873c2f1-9aab-4cd8-a7b5-ccbb085b679d.png)

But, if implemented, I'm once again in doubt as to when the progress bar should be shown. It won't show up if we pass `-s`, but should we show the bar always, only when -v is used or only when there are many files?
I have a feeling it should always be shown, but I'm no expert in UI/UX design

deseven commented:
It's alright, I also want to make a good solution for uploading :)

My original idea was:
1. Default mode (without any arguments) - show all steps and progress bars where needed, print summary in the end.
2. Verbose mode (with `-v`) - show all steps, filenames instead of progress bars and more diagnostic information where applicable, print summary in the end.
3. Silent mode (with `-s`) - print only summary in the end.

Kinda like `ping` with addition of either `-q` or `-s` flags.

Don't worry too much about making the progress bar for threaded uploading, I think I'll be able to add it afterwards.

Commit:
Prettier MD5 checks

Display the progress bar when performing MD5 checks, letting the user
know what's going on after pressing 'U' to upload files. Also, if
verbose, display the current working directory first and then just the
file names, to not clutter the terminal, and inform the user about
skipped files as soon as they're checked against the server.

Sonico98 commented:
I think that's it. Please let me know your opinion when you can test it.
Also please tell me if you would like me to squash the commits, or have some ideas for code improvements :)
And getting the progress bar to work would be great!

deseven commented on ibroadcast-uploader.py:
```diff
@@ -287,35 +277,94 @@ def calcmd5(self, filePath="."):
-                skipped += 1
-                continue
+                self.skipped_files.append(filename)
+                if not self.be_silent and self.be_verbose and calculated:
```
What if the file checksum wasn't calculated (because we already have it)? In verbose mode we should still display the info that we're skipping the file because it was already uploaded, no matter whether we calculated the checksum or not.

Not sure I understand why this `calculated` flag is needed at all. Please clarify.

Sonico98 replied:
My bad, I added that check because otherwise simply the text "Skipping, already uploaded" would be shown. For some reason I didn't realize at the time that yes, that information should be printed if we're in verbose mode. I'll rework it.

deseven commented on ibroadcast-uploader.py:
```diff
@@ -287,35 +277,94 @@ def calcmd5(self, filePath="."):
+            print("\n", flush=True, file=out)
+
+    def split_file_list(self, slices):
+        self.files = [self.files[i * slices:(i + 1) * slices] for i in range((len(self.files) + slices - 1) // slices )]
```
What this does and why it's being called just once with `slices` = 1? I think this piece of code needs a comment.

Sonico98 replied:
What this does is split the list into chunks of size `slices`, so when we pass the list to the `upload()` function, only a single file is taken per thread (which is why we call it with the value of 1). I had tried avoiding doing this, but encountered the problem that the whole string was being printed out on screen, letter by letter. And only now I realize what the problem was: I was later iterating over the string inside the `upload()` function; I just had to remove the for loop and reindent the remaining code.

Commit:
Remove unnecesary code, correct verbose printing

Splitting the list into chunks of just one element is not necessary, as
we can just access its contents one by one.

Correctly inform the user what's happening during MD5 generation if the
`-v` flag is passed:
- Always print the file name of the current element
- Inform any of the following situations:
    - Calculating the file's MD5 hash
    - The file's hash is cached, but the file has not been uploaded yet
    - Skipping the file if it's already been uploaded, regardless if we
      calculated the hash or not

deseven commented:
I'm not really happy with vars such as `print_filename_again` and `not_skipped_files`, I think that the first one can be avoided and the second one is just confusing. But this is not important :)

I'm going to merge this now, it seems to be working just fine (tested on the latest macOS and Debian) 👍
I might alter some stuff later and restore the progress bar, don't have any spare time this week.

Thanks a lot!

Sonico98 commented:
Thanks! I'll be waiting for your changes, specially the progress bar.

deseven commented:
Still have absolutely no time to work on that, I guess we'll stay without progress bar for some time :) I have it in my to do, though.
